pkg/arvo/neo/cod/std/src/fil/tasks.md

```diff
@@ -0,0 +1,624 @@
+
+In this lesson we’ll see how shrubs keep their parents informed of state changes using the `%gift` poke. This is the most complex UI we’ve looked at yet, so we’ll also focus on the `/con` files.
+
+## /imp/task.hoon
+Tasks only needs one `/imp` file: `task.hoon`. The Tasks frontend shows you some tasks that may or may not have other tasks as children.
+
+```hoon
+/@  task  ::  [text=cord done=? order=(list path]
+::
+::  $task-diff
+::  $%  [%new =task prepend=?]
+::      [%edit text=cord done=?]
+::      [%oust =pith]
+::      [%reorder order=(list pith)]
+::  ==
+/@  task-diff
+=>
+::
+::  helper core
+|%
+::
+::  check if all kids are completed
+++  check-kids
+  |=  =bowl:neo
+  ^-  ?
+  ?:  =([~ ~] kids.bowl)
+    %.y
+  =/  piths  ~(tap in ~(key by ~(tar of:neo kids.bowl)))
+  %+  levy  piths
+  |=  =pith
+  =/  =task
+    !<  task
+    q.pail:(need (~(get by ~(tar of:neo kids.bowl)) pith))
+  done.task
+::
+::  assign a unique numerical ID to a new subtask
+++  assign-name
+  |=  =bowl:neo
+  ^-  @ud
+  ?:  =([~ ~] kids.bowl)  1
+  =/  sorted-names=(list @ud)
+    %-  sort  :_  lth
+    %+  turn  ~(tap by ~(tar of:neo kids.bowl))
+      |=  [=pith =idea:neo]
+      +:(,[%ud @ud] (rear pith))
+  =/  last-name=@ud  (rear sorted-names)
+  =/  name-missing=(list @ud)
+    %+  skim  (gulf 1 last-name)
+    |=  n=@ud
+    =(~ (find ~[n] sorted-names))
+  ?~  name-missing  +(last-name)
+  (rear name-missing)
+--
+::
+::  outer core
+^-  kook:neo
+|%
+::
+::  state is a %task
+++  state  pro/%task
+::
+::  we accept %gift and %task-diff pokes
+++  poke   (sy %task-diff %gift ~)
+::
+::  we define one generation of kids at
+::  /path/to/this/task/<@ud> which are
+::  also tasks like their parent, which
+::  means those kids also define one
+::  generation of kids below them, and so on
+++  kids
+  :+  ~  %y
+  %-  ~(gas by *lads:neo)
+  :~  :-  [|/%ud |]
+    [pro/%task (sy %task-diff %gift ~)]
+  ==
+++  deps
+  ::
+  ::  we don't need dependencies to negotiate
+  ::  state between parent and children tasks,
+  ::  but it would be valid to do so
+  *deps:neo
+++  form
+  ::
+  ::  inner core
+  ^-  form:neo
+  ::  XX refactor sample to same as other tutorials
+  |_  [=bowl:neo =aeon:neo stud:neo state-vase=vase]
+  ::  XX refactor this to default +init from counter
+  ++  init
+    |=  pal=(unit pail:neo)
+    ^-  (quip card:neo pail:neo)
+    `(need pal)
+  ++  poke
+    |=  [=stud:neo vax=vase]
+    ^-  (quip card:neo pail:neo)
+    =/  this  !<(task state-vase)
+    ?+    stud  !!
+        %gift
+      ?:  (check-kids bowl)
+        [~ task/!>(this(done %.y))]
+      [~ task/!>(this(done %.n))]
+    ::
+        %task-diff
+      =/  diff  !<(task-diff vax)
+      ?-    -.diff
+          %new
+        =/  name=@ud  (assign-name bowl)
+        =.  order.this
+          ?:  prepend.diff
+            ::  XX do we really need to do this? bad look if so
```

Comment:
Are you referring to the casting? I agree it's messy, let's try it without more often.